### PR TITLE
Add shuffling in Nanotron for subsequent epochs when data is repeated

### PR DESCRIPTION
**Pull request by Lauler**
Nanoset's index builder does not re-shuffle dataset and sample indices within epochs when training secondary, third, etc epochs. It instead concatenates a copy of the same indices for any repeated data. This PR adds unique within-epoch shuffling for each epoch.

See the following issue: https://github.com/huggingface/nanotron/issues/237 

I ran the tests in `tests/nanotron`:

```
=========================================================================== warnings summary ===========================================================================
tests/helpers/context.py:7: 35 warnings
  /home/faton/projects/text/nanotron_dev/nanotron/tests/helpers/context.py:7: PytestCollectionWarning: cannot collect test class 'TestContext' because it has a __init__ constructor (from: nanoset/test_build_nanoset_dataloader.py)
    class TestContext:

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================== 12 passed, 35 warnings in 44.55s ===================================================================
```

### CODE REVIEW

File: src/nanotron/data/nanoset.py [hunk 1]
```diff
@@ -111,14 +111,26 @@ def build_nanoset_index(self) -> np.ndarray:
...
-        numpy_random_state.shuffle(dataset_index)
-        numpy_random_state = np.random.RandomState(self.random_seed)
-        numpy_random_state.shuffle(dataset_sample_index)
-        # Concatenate num_epochs the shuffled indexes
-        dataset_index = np.concatenate([dataset_index for _ in range(num_epochs)])
-        dataset_sample_index = np.concatenate([dataset_sample_index for _ in range(num_epochs)])
+
+        # Shuffle indices in each epoch with different random seeds and concatenate them
+        r = np.random.RandomState(self.random_seed)
+        epoch_random_seeds = r.randint(0, 2**32 - 1, num_epochs)
```
Comment (NouamaneTazi) [inline src/nanotron/data/nanoset.py:117]: Can we just use `self.random_seed + num_epoch` for easier reproducibility?

**Review by NouamaneTazi — APPROVED** [review]
Nice catch! LGTM 🤗

File: src/nanotron/data/nanoset.py (hunk 1 above)
Comment (Lauler) [reply src/nanotron/data/nanoset.py]: Good suggestion, makes it a lot simpler and clearer what's going on. I've changed the code to incorporate this in latest commit:

https://github.com/huggingface/nanotron/blob/f060414194e02c2e22e8fa0ffd8326363917a796/src/nanotron/data/nanoset.py#L116-L123
